### PR TITLE
chore(aft): Added generate amplify-swift

### PR DESCRIPTION
**Pull request by Equartey**
*Description of changes:*
Adds `aft generate amplify-swift` command
- `-b` to change target branch
- `-o` to change output
- `-d` to run a diff of current vs incoming
- Sets cloned files to read-execute. This prevents modifications, but allows `pod install`
- Added internal definitions for pods

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+  /// Sets up the Amplify Swift repo for use later
+  Future<void> _setupRepo() async {
+    if (_repoCache[branchTarget] != null) {
+      return;
+    }
+    Directory repoDir;
+    repoDir = await _downloadRepository();
+    repoDir = await _checkoutRepositoryRef(repoDir, branchTarget);
+
+    _repoCache[branchTarget] = repoDir;
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:151]: Suggestion: You can make `repoDir` final if you refactor this a bit

```suggestion
    final repoDir = await _checkoutRepositoryRef(await _downloadRepository(), branchTarget);

    _repoCache[branchTarget] = repoDir;
```

Comment (Jordan-Nelson) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:151]: Alternatively, something like this would work as well.

```suggestion
    final repoDir = await _downloadRepository();
    final repoRef = await _checkoutRepositoryRef(repoDir, branchTarget);

    _repoCache[branchTarget] = repoRef;
```

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+    final outputDir = Directory(p.join(outputPath, plugin.name));
+
+    // Clear out the directory if it already exists.
+    // This is to ensure that we don't have any stale files.
+    if (await outputDir.exists()) {
+      logger.verbose(
+        'Clearing out existing plugin directory for ${plugin.name}...',
+      );
+      await _setPermissions(outputDir.path, PermissionScope.readExecuteWrite);
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:180]: Q: Do you need to set permissions to execute if you are just removing the files?

Comment (Jordan-Nelson) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:180]: Q: What permissions exist without this?

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+      );
+    }
+    logger.info(
+      'No differences between incoming and current for ${plugin.name}.',
+    );
+  }
+
+  /// Sets the permissions of all files in [path]  to
+  Future<void> _setPermissions(String path, PermissionScope scope) async {
+    const scope = '777';
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:223]: Q: Does this need to be updated?

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      ..info('Copying plugin files for ${plugin.name}...')
+      ..verbose('From $repoDir to $outputDir');
+    await copyPath(repoDir.path, outputDir.path);
+    await _setPermissions(outputDir.path, PermissionScope.readExecute);
+  }
+
+  Future<void> checkDiff(PluginConfig plugin) async {
+    logger.info('Checking diff for ${plugin.name}...');
+    final incoming = (await _pluginDirForPath(plugin.remotePathToSource)).path;
+    // Set the permissions of the incoming files to readExecute (fails otherwise)
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:196]: Q: What fails when the permissions are not updated?

File: packages/aft/lib/src/util.dart [hunk 5]
```diff
@@ -115,3 +116,34 @@ void sortPackagesTopologically<T>(
         .compareTo(ordered.indexOf(getPubspec(a).name));
   });
 }
+
+// Copyright 2017, the Dart project authors.
+// Copied from io.dart https://pub.dev/documentation/io/latest/io/copyPath.html
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/util.dart:121]: Q: Is this comment/link correct?

File: packages/aft/lib/src/util.dart [hunk 6]
```diff
@@ -115,3 +116,34 @@ void sortPackagesTopologically<T>(
...
+  }
+  if (p.isWithin(from, to)) {
+    throw ArgumentError('Cannot copy from $from to $to');
+  }
+  return false;
+}
+
+// Copyright 2017, the Dart project authors.
+// Copied from io.dart https://pub.dev/documentation/io/latest/io/copyPath.html
+Future<void> copyPath(String from, String to) async {
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/util.dart:134]: Q: Why not use the copyPath method from Dart io?

File: packages/aft/lib/src/util.dart [hunk 7]
```diff
@@ -115,3 +116,34 @@ void sortPackagesTopologically<T>(
...
+  if (p.canonicalize(from) == p.canonicalize(to)) {
+    return true;
+  }
+  if (p.isWithin(from, to)) {
+    throw ArgumentError('Cannot copy from $from to $to');
+  }
+  return false;
+}
+
+// Copyright 2017, the Dart project authors.
```
Comment (Jordan-Nelson) [inline packages/aft/lib/src/util.dart:132]: Q: Is this a standard way to provide attribution?

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart (hunk 1 above)
Comment (Equartey) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:151]: Good suggestion, took the later.

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart (hunk 4 above)
Comment (Equartey) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:196]: diff command catches the literal difference of permissions. Ie, 555 local vs 755 remote.

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart (hunk 3 above)
Comment (Equartey) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:223]: Good catch, will remove.

File: packages/aft/lib/src/util.dart (hunk 5 above)
Comment (Equartey) [reply packages/aft/lib/src/util.dart:121]: Discussed offline: Added io package as a dep.

File: packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart (hunk 2 above)
Comment (Equartey) [reply packages/aft/lib/src/commands/generate/generate_amplify_swift_command.dart:180]: After looking into this more, turns out git doesn't preserve file permissions. Will remove this logic.

**Review by khatruong2009 — APPROVED** [review]

**Review by khatruong2009 — APPROVED** [review]

File: packages/amplify_datastore/example/ios/Podfile [hunk 8]
```diff
@@ -31,6 +31,11 @@ target 'Runner' do
   use_frameworks!
   use_modular_headers!
 
+  # Point pods to internal source files
```
Comment (Jordan-Nelson) [inline packages/amplify_datastore/example/ios/Podfile:34]: What is the purpose of these changes? Will customers have to make changes in their app?

**Review by Jordan-Nelson — APPROVED** [review]
